spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/GraphGeneratorVisitor.java

```diff
@@ -0,0 +1,651 @@
+
+		// Transitions made from inside this sequence which are not satisfied
+		final List<Context.TransitionTarget> outstandingTransitions = new ArrayList<>();
+		
+		Flow primaryFlow;
+		
+		final Map<Flow,Map<String,Node>> labeledNodesInEachFlow = new HashMap<>();
+		
+		public Sequence(int sequenceNumber, String label, Node sequenceStartNode) {
+			this.sequenceNumber = sequenceNumber;
+			this.label = label;
+			nodes.add(sequenceStartNode);
+		}
+
+		public String toString() {
+			StringBuilder s = new StringBuilder();
+			for (Node n : nodes) {
+				s.append("[").append(n.id).append(":").append(n.name).append("]");
+			}
+			for (Link l : links) {
+				s.append("[" + (l.getTransitionName() == null ? "" : l.getTransitionName() + ":") + l.from + "-" + l.to
+						+ "]");
+			}
+			s.append("  transitions:").append(outstandingTransitions);
+			s.append("   flowLabelsMap:").append(labeledNodesInEachFlow);
+			return s.toString();
+		}
+
+	}
+
+	static class Context {
+
+		// Id of the first node in this context (start of a flow, start of a
+		// split)
+		private String startNodeId;
+
+		// As a flow/split is processed gradually we accumulate dangling nodes,
+		// those that need connecting to whatever comes next. For a flow there might
+		// just be one, but for a split multiple.
+		private List<String> currentDanglingNodes = new ArrayList<>();
+
+		// When a transition branch is taken to some other app within a flow,
+		// this records the app node that must be joined to the exit path 
+		// from the entire flow.
+		public List<String> otherExits = new ArrayList<>();
+
+		// Within a flow, transitions to the 'same job' would share a node
+		// target, as would all transitions to an $END or $FAIL node. This 
+		// keeps track of what has been created so it can be reused.
+		Map<String, Node> nodesSharedInFlow = new LinkedHashMap<String, Node>();
+
+		// When processing apps in a real Flow or Split, this is the Ast node
+		// for that Flow or Split.
+		LabelledComposedTaskNode containingNode;
+
+		// Nodes in this sequence that are labeled are recorded here
+		final Map<String, Node> nodesWithLabels = new HashMap<>();
+
+		// Tracking what kind of context we are in
+		boolean isFlow = false;
+		boolean isSplit = false;
+
+		// For forward references this keeps track of them so that they can be
+		// filled in later. The reference must be satisfied within the same flow
+		// or by a separate sequence. If that secondary sequence routes back 
+		// to the primary, it must be within the same flow!
+		public List<Context.TransitionTarget> transitionTargets = new ArrayList<>();
+
+		public Map<String, Node> extraNodes = new HashMap<>();
+
+		Context(boolean isFlow, boolean isSplit, String startNodeId, LabelledComposedTaskNode split) {
+			this.isFlow = isFlow;
+			this.isSplit = isSplit;
+			this.startNodeId = startNodeId;
+			this.containingNode = split;
+		}
+
+		public void addDanglingNodes(boolean replace, String... is) {
```

Comment:
rename replace with replaceExisting to be consistent with addDanglingNodes below.